File: python/12.整数转罗马数字.py
#
# @lc app=leetcode.cn id=12 lang=python
#
# [12] 整数转罗马数字
#
# 解法1(T88% S48%)：没什么技巧的纯流程题，按照题干依次判断，首先先while减去足量的1000，然后判断剩余是否大于900/大于500/大于400，最后再while减去足量的100；然后再判断剩余是否大于90/50/40，然后再while减去足量的10；最后再判断剩余是否大于9/5/4，最终再减去足量的1直到num==0

# @lc code=start
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        roman = ""
        while num >= 1000:
            roman += 'M'
            num -= 1000
        if num >= 900:
            roman += 'CM'
            num -= 900
        elif num >= 500:
            roman += 'D'
            num -= 500
        elif num >= 400:
            roman += 'CD'
            num -= 400
        while num >= 100:
            roman += 'C'
            num -= 100
        
        if num >= 90:
            roman += 'XC'
            num -= 90
        elif num >= 50:
            roman += 'L'
            num -= 50
        elif num >= 40:
            roman += 'XL'
            num -= 40
        while num >= 10:
            roman += 'X'
            num -= 10
        
        if num >= 9:
            roman += 'IX'
            num -= 9
        elif num >= 5:
            roman += 'V'
            num -= 5
        elif num >= 4:
            roman += 'IV'
            num -= 4
        while num:
            roman += 'I'
            num -= 1
        
        return roman
# @lc code=end

File: javascript/12.整数转罗马数字.js
/*
 * @lc app=leetcode.cn id=12 lang=javascript
 *
 * [12] 整数转罗马数字
 */

// @lc code=start
/**
 * @param {number} num
 * @return {string}
 */
var intToRoman = function(num) {
    var res = "";

    while (num >= 1000) {
        res = res + "M";
        num -= 1000;
    }
    if (num >= 900) {
        res = res + "CM";
        num -= 900;
    }
    if (num >= 500) {
        res = res + "D";
        num -= 500;
    }
    if (num >= 400) {
        res = res + "CD";
        num -= 400;
    }
    while (num >= 100) {
        res = res + "C";
        num -= 100;
    }
    if (num >= 90) {
        res = res + "XC";
        num -= 90;
    }
    if (num >= 50) {
        res = res + "L";
        num -= 50;
    }
    if (num >= 40) {
        res = res + "XL";
        num -= 40;
    }
    while (num >= 10) {
        res = res + "X";
        num -= 10;
    }
    if (num >= 9) {
        res = res + "IX";
        num -= 9;
    }
    if (num >= 5) {
        res = res + "V";
        num -= 5;
    }
    if (num >= 4) {
        res = res + "IV";
        num -= 4;
    }
    while (num) {
        res = res + "I";
        num -= 1;
    }
    return res;
};
// @lc code=end

